### ISSUE TITLE
Possible leak with Local

### ISSUE THREAD

**Issue by alexandru** [monix/monix#1023]
Reported on Gitter, see [September 10, 2019 10:45 AM](https://gitter.im/monix/monix?at=5d7754a11882325229974e05).

Reproducible example: https://gist.github.com/hanny24/3096f1541e1f05716f8c477691c36f2f

To be investigated.

**Comment by Avasil** [comment]
@hanny24 
Which version of http4s do you use? I'm running it for 30 mins on jdk1.8 with http4s 0.20.10 and so far I haven't noticed any indication of memory leak. I will leave it running for the remainder of the night. How long does it take to manifest?

**Comment by hanny24** [comment]
I'm using monix RC5, http4s 0.20.10, cats effect 1.4.0 and OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.3+7). I'm running it JVM options `-Xmx64m -XX:+HeapDumpOnOutOfMemoryError`.  

I'm using `ab` to generate a lot of traffic: `ab -k -n 50000 -c 10 http://127.0.0.1:8080/hello` and it usually takes just a few seconds to run out of heap space. 

I can even provide you with a heap dump. Available on my [gdrive](https://drive.google.com/file/d/1z-GvVNlIIjdwjOHoYZb2VnWexp1UWIFt/view?usp=sharing).

Quick check in Memory Analyzer shows this:
```
The thread scala.concurrent.impl.ExecutionContextImpl$DefaultThreadFactory$$anon$1 @ 0x7fc4fd278 scala-execution-context-global-16 keeps local variables with total size 25 732 920 (39,43 %) bytes.

The memory is accumulated in one instance of "java.lang.ThreadLocal$ThreadLocalMap$Entry[]" loaded by "<system class loader>".
The stacktrace of this Thread is available. See stacktrace.

Keywords
java.lang.ThreadLocal$ThreadLocalMap$Entry[]
```

Let me know if I can provide any further assistance.

**Comment by hanny24** [comment]
Example SBT project: https://github.com/hanny24/monix-bug

**Comment by guymers** [comment]
Could this be because the thread pool is unbounded and your computer cannot process the volume of requests?
If you used a fixed size pool with a fixed queue size is memory usage constant?
Alternatively if you bench with a much smaller number of requests such that that your computer can process them without needing to queue is memory usage constant?

**Comment by hanny24** [comment]
Both `Scheduler.traced` and `Scheduler.global` are using the same thread pool.

**Comment by Avasil** [comment]
@guymers It's certainly a memory leak but it does not happen on JDK 8 and I feel like it's not related to `Local`.

I think the problem happens in this line: https://github.com/monix/monix/blob/master/monix-eval/shared/src/main/scala/monix/eval/Task.scala#L4376

Even if it's not a `TracingScheduler` but something else and locals are completely disabled, I encounter the leak on Java 11. For some reason the wrappers and associated structures aren't cleaned up. So far I wasn't able to reproduce it without http4s so there is probably some weird interaction going on. 

I haven't managed to figure out the root cause yet. I need to deal with few lower hanging fruits on my backlog and then I'll try again, perhaps with fresh ideas

**Comment by alexandru** [comment]
We expose the `Scheduler` in builders such as `deferAction` and `deferFutureAction` among others. That `Scheduler` reference must be captured somehow in a way that's never released.

It's a really odd problem. But it could also be http4s related.

**Comment by Avasil** [comment]
My suspicion is that it would also leak for `cats.effect.IO` if there was any wrapping but haven't tried it yet

**Comment by alexandru** [comment]
It's a different implementation though. `cats.effect.IO` does not have an underlying `ExecutionContext`, the only way you can talk about such a thing is with `ContextShift`, although building a `ContextShift` instance that's always based on a "wrapped" `global` would be worth trying, although I'm not sure it will accomplish anything, since it's a little apples vs oranges.

**Comment by Avasil** [comment]
@hanny24 The issue should be fixed in `3.1.0` - please let us know if it's not the case!
